sycl/include/sycl/accessor.hpp

```diff
@@ -2646,21 +2646,18 @@ class __SYCL_SPECIAL_CLASS local_accessor_base :
     return detail::convertToArrayOfN<Dims, 1>(getSize());
   }
 
-  template <int Dims = Dimensions,
-            typename = std::enable_if_t<Dims == 0 && IsAccessAnyWrite>>
+  template <int Dims = Dimensions, typename = std::enable_if_t<Dims == 0>>
   operator RefType() const {
     return *getQualifiedPtr();
   }
 
-  template <int Dims = Dimensions,
-            typename = std::enable_if_t<(Dims > 0) && IsAccessAnyWrite>>
+  template <int Dims = Dimensions, typename = std::enable_if_t<(Dims > 0)>>
   RefType operator[](id<Dimensions> Index) const {
     const size_t LinearIndex = getLinearIndex(Index);
     return getQualifiedPtr()[LinearIndex];
   }
 
-  template <int Dims = Dimensions,
-            typename = std::enable_if_t<Dims == 1 && IsAccessAnyWrite>>
+  template <int Dims = Dimensions, typename = std::enable_if_t<Dims == 1>>
```

Comment:
Based on the failures in CI, I believe you might have to add `&& AccessMode != access::mode::atomic` in here, as otherwise it conflicts with the below definitions.

Tag @gmlueck for awareness. The spec doesn't make any special cases for `local_accessor` with the deprecated `access_mode::atomic`. Should we worry about the return type for that access mode, or do we have enough wiggle-room to keep it returning `atomic` for said access mode?

Reply:
Thanks @steffenlarsen, I added IsAccessReadOnly to initial check `IsAccessAnyWrite`. I was thinking of your suggestion, but this sound a bit more clear on what needed to satisfy the condition.

Reply:
> Tag @gmlueck for awareness. The spec doesn't make any special cases for local_accessor with the deprecated access_mode::atomic. Should we worry about the return type for that access mode, or do we have enough wiggle-room to keep it returning atomic for said access mode?

If we decide we need this changed, we can address it in a follow-up.

Reply:
I think this part of the code is implementing the deprecated `target::local` for `accessor`, not the `local_accessor` class.  Right?

The spec does exclude `access_mode::atomic` in this case.  [Section 4.7.6.9.4.7](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_accessor_specialization_with_targetlocal) says that this specialization of `accessor` is only available for `access_mode:read_write` and `access_mode::atomic`:

> This specialization of accessor is only available for access modes `access_mode::read_write` and `access_mode::atomic`.

And the synopsis says this form of `operator[]` is only available for `access_mode::read_write`, which excludes `access_mode::atomic`:

```
/* Available only when: (AccessMode == access_mode::read_write && Dimensions
  * == 1) */
  reference operator[](size_t index) const;
```

From what I can tell, your implementation also excludes `access_mode::atomic` because `(IsAccessAnyWrite || IsAccessReadOnly)` does not include that access mode.

Reply:
This is defined in a common base-class shared between both `local_accessor` and `accessor` with `target::local`. If we wanted them to differ we could just move the definition.

Reply:
Note that `local_accessor` does not even have a template parameter for `access_mode`.

Reply:
> Note that `local_accessor` does not even have a template parameter for `access_mode`.

Ah, sorry. In that case I believe everything should be alright.